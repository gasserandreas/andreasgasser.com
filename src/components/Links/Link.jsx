import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { Link as GLink } from 'gatsby';

const getLinkStyles = ({ theme }) => `
  color: ${theme.colors.textHighlighed};
  font-weight: 600;
  font-size: 0.95rem;
  text-decoration: none;
  text-transform: uppercase;

  &:hover {
    color: ${theme.colors.textHover};
  }
`;

const Styles = {
  GLink: styled(GLink)((props) => `
    ${getLinkStyles(props)};
  `),
  ExternalLink: styled.a((props) => `
    ${getLinkStyles(props)};
  `),
};

const Link = ({
  children,
  to,
  external,
  ...props
}) => {
  if (external) {
    return (
      <Styles.ExternalLink
        href={to}
        target="_blank"
        {...props}
      >
        {children}
      </Styles.ExternalLink>
    );
  }

  return (
    <Styles.GLink
      to={to}
      {...props}
    >
      {children}
    </Styles.GLink>
  );
};

export const propTypes = {
  children: PropTypes.node.isRequired,
  to: PropTypes.string.isRequired,
  external: PropTypes.bool,
};

export const defaultProps = {
  external: false,
};

Link.propTypes = propTypes;

Link.defaultProps = defaultProps;

export default Link;
