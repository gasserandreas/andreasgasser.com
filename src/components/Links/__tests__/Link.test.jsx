/* global withTheme */
import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Link from '../Link';

describe('components/Link', () => {
  const defaultProps = {
    to: 'https://www.google.com/',
    external: true,
    children: 'click me',
  };

  it('should render a tag with target new prop for external links', () => {
    const dom = mount(withTheme(<Link {...defaultProps} />));
    expect(dom.exists()).toBeTruthy();

    // should not pass Link props
    const props = dom.props();
    expect(props.to).toBeUndefined();

    expect(dom.getDOMNode()).toHaveProperty('target', '_blank');
    expect(dom.getDOMNode()).toHaveProperty('href', defaultProps.to);
  });

  it('simple snapshot test', () => {
    const dom = mount(withTheme(<Link {...defaultProps} />));
    expect(toJson(dom)).toMatchSnapshot();
  });
});
