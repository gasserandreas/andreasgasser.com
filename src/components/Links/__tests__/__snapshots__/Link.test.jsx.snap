// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Link simple snapshot test 1`] = `
<CustomThemeProvider
  dark={false}
>
  <ThemeProvider
    theme={
      Object {
        "breakpoints": Array [
          "320px",
          "540px",
          "720px",
          "960px",
          "1140px",
          "1400px",
        ],
        "buttons": Object {
          "outline": Object {
            "bg": "transparent",
            "boxShadow": "inset 0 0 2px",
            "color": "primary",
            "variant": "buttons.primary",
          },
          "primary": Object {
            "bg": "primary",
            "borderRadius": "default",
            "color": "background",
            "fontSize": 2,
            "fontWeight": "bold",
          },
          "secondary": Object {
            "bg": "secondary",
            "color": "background",
            "variant": "buttons.primary",
          },
        },
        "colors": Object {
          "background": "#fff",
          "gray": "#dddddf",
          "highlight": "hsla(205, 100%, 40%, 0.125)",
          "muted": "#f6f6f9",
          "primary": "#07c",
          "secondary": "#30c",
          "text": "#000",
        },
        "fontSizes": Array [
          14,
          16,
          20,
          24,
          32,
          48,
          64,
          96,
        ],
        "fontWeights": Object {
          "body": 400,
          "bold": 900,
          "heading": 700,
        },
        "fonts": Object {
          "body": "Open Sans, Helvetica Neue, Helvetica, sans-serif",
          "heading": "inherit",
          "monospace": "Menlo, monospace",
        },
        "lineHeights": Object {
          "body": 1.5,
          "heading": 1.125,
        },
        "radii": Object {
          "circle": 99999,
          "default": 4,
        },
        "shadows": Object {
          "card": "0 0 4px rgba(0, 0, 0, .125)",
        },
        "sizes": Object {
          "avatar": 48,
        },
        "space": Array [
          0,
          4,
          8,
          16,
          32,
          64,
          128,
          256,
          512,
        ],
        "styles": Object {
          "root": Object {
            "fontFamily": "body",
            "fontWeight": "body",
            "lineHeight": "body",
          },
        },
        "text": Object {
          "bold": Object {
            "color": "#595959",
            "fontWeight": "900",
          },
          "default": Object {
            "color": "#595959",
          },
          "muted": Object {
            "color": "#808080",
          },
          "page": Object {
            "fontSize": 4,
            "letterSpacing": "-0.04em",
            "lineHeight": "1.4em",
            "wordSpacing": "-0.08em",
          },
          "section": Object {
            "fontSize": 3,
            "lineHeight": "1.4em",
          },
          "small": Object {
            "color": "#595959",
            "fontSize": 0,
          },
          "subPage": Object {
            "color": "#808080",
            "fontSize": 3,
            "lineHeight": "1.6em",
          },
        },
        "variants": Object {
          "avatar": Object {
            "borderRadius": "circle",
            "height": "avatar",
            "width": "avatar",
          },
          "card": Object {
            "bg": "background",
            "boxShadow": "card",
            "p": 2,
          },
          "link": Object {
            "color": "primary",
          },
          "nav": Object {
            ":hover,:focus,.active": Object {
              "color": "primary",
            },
            "color": "inherit",
            "display": "inline-block",
            "fontSize": 1,
            "fontWeight": "bold",
            "p": 2,
            "textDecoration": "none",
          },
        },
      }
    }
  >
    <Link
      external={true}
      to="https://www.google.com/"
    >
      <styled.a
        href="https://www.google.com/"
        target="_blank"
      >
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bwzfXH",
                "isStatic": false,
                "lastClassName": "jbnfhe",
                "rules": Array [
                  [Function],
                ],
              },
              "displayName": "styled.a",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bwzfXH",
              "target": "a",
              "toString": [Function],
              "usesTheme": false,
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          href="https://www.google.com/"
          target="_blank"
        >
          <a
            className="sc-bwzfXH jbnfhe"
            href="https://www.google.com/"
            target="_blank"
          >
            click me
          </a>
        </StyledComponent>
      </styled.a>
    </Link>
  </ThemeProvider>
</CustomThemeProvider>
`;
